.github/scripts/compile_check/compile_check.py

```diff
@@ -0,0 +1,388 @@
+#!/usr/bin/env python3
+"""Compile and dependency validation tool for Kubeflow Pipelines components.
+
+This script discovers component and pipeline modules based on the presence of
+`metadata.yaml` files, validates declared dependencies, and ensures each target
+compiles successfully with the Kubeflow Pipelines SDK.
+"""
+
+from __future__ import annotations
+
+import argparse
+import importlib
+import logging
+import sys
+import tempfile
+import traceback
+from dataclasses import dataclass, field
+from pathlib import Path
+from typing import Dict, Iterable, List, Optional, Sequence, Tuple
+
+import yaml
+
+try:
+    from packaging.specifiers import SpecifierSet
+except ImportError:  # pragma: no cover - packaging is optional
+    SpecifierSet = None  # type: ignore[assignment]
+
+from kfp import compiler as pipeline_compiler
+from kfp.dsl import base_component, graph_component
+
+REPO_ROOT = Path(__file__).resolve().parents[1]
+
+
+@dataclass
+class MetadataTarget:
+    """Represents a single component or pipeline discovered from metadata."""
+
+    metadata_path: Path
+    module_path: Path
+    module_import: str
+    tier: str
+    target_kind: str  # "component" or "pipeline"
+    metadata: Dict
+
+
+@dataclass
+class ValidationResult:
+    """Stores the outcome of validating a single metadata target."""
+
+    target: MetadataTarget
+    success: bool
+    compiled_objects: List[str] = field(default_factory=list)
+    warnings: List[str] = field(default_factory=list)
+    errors: List[str] = field(default_factory=list)
+
+    def add_error(self, message: str) -> None:
+        """Record a validation error and mark the result unsuccessful."""
+        logging.error(message)
+        self.errors.append(message)
+        self.success = False
+
+    def add_warning(self, message: str) -> None:
+        """Record a non-fatal validation warning."""
+        logging.warning(message)
+        self.warnings.append(message)
+
+
+def parse_args(argv: Optional[Sequence[str]] = None) -> argparse.Namespace:
+    """Parse command-line arguments for the compile check tool."""
+    parser = argparse.ArgumentParser(description="Compile Kubeflow components and pipelines.")
+    parser.add_argument(
+        "--tier",
```

Comment:
I think we can remove the tier filtering since we are not going to have third party components or pipelines.